Integration/Load test framework

pranavbhandari24 opened the pull request:
Integration / Load test utilities.
------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [x] Mention the appropriate issue in your description (for example: `addresses #123`), if applicable. This will automatically add a link to the pull request in the issue. If you would like the issue to automatically close on merging the pull request, comment `fixes #<ISSUE NUMBER>` instead.
 - [x] Update `CHANGES.md` with noteworthy changes.
 - [x] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/get-started-contributing/#make-the-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/workflows/Go%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

kennknowles commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/matchers/CassandraAsserts.java:
```diff
@@ -0,0 +1,80 @@
+   * @param rows Rows to parse.
+   * @return List of maps to use in {@link RecordsSubject}.
+   */
+  @SuppressWarnings("argument")
```
why suppressed?

pranavbhandari24 replied:
I have disabled checker framework on all `it` submodules. As suggested, I created an issue to enable checker framework on it submodules after the code drop.

pranavbhandari24 replied:
https://github.com/apache/beam/issues/27438

kennknowles commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/DefaultCassandraResourceManager.java:
```diff
@@ -0,0 +1,276 @@
+  public static final class Builder
+      extends TestContainerResourceManager.Builder<DefaultCassandraResourceManager> {
+
+    @SuppressWarnings("initialization.fields.uninitialized")
```
why suppressed?

kennknowles commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/DefaultCassandraResourceManager.java:
```diff
@@ -0,0 +1,276 @@
+        builder);
+  }
+
+  @SuppressWarnings({"rawtypes", "method.invocation"})
```
why suppressed?

pranavbhandari24 replied:
Will address after code drop. Opened https://github.com/apache/beam/issues/27438.

kennknowles reviewed (changes requested):
This is a pretty huge code drop. I am just partway through the files. But I thought I would just make a top level comment that it would be better to not suppress the warnings, but to fix them. If they are left in place, they should have a reason or at least a link to a GitHub Issue something like "Fix suppressed errors after integration test framework code drop"

kennknowles commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/CassandraResourceManagerUtils.java:
```diff
@@ -0,0 +1,52 @@
+
+import static org.apache.beam.it.common.utils.ResourceManagerUtils.generateResourceId;
+
+import com.google.re2j.Pattern;
```
Why this over the stuff that ships with Java? It seems basically the same

pranavbhandari24 replied:
Changed to use java Patterns in https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/710.

kennknowles commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/DefaultCassandraResourceManager.java:
```diff
@@ -0,0 +1,276 @@
+import java.net.InetSocketAddress;
+import java.util.List;
+import java.util.Map;
+import javax.annotation.Nullable;
```
Prefer checkerframework annotations since they can go in more places (like on type uses)

pranavbhandari24 replied:
I have kept the java annotations for now since we don't need the checker annotations yet. We can decide on this after the code drop.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/ResourceManager.java:
```diff
@@ -0,0 +1,25 @@
+package org.apache.beam.it.common;
+
+/** Common interface across resource managers. */
+public interface ResourceManager {
```
This seems close to `Closeable` or `Autocloseable`. Would it make sense to use it in that style?

pranavbhandari24 replied:
Changed the resource managers to match `Closable` and `AutoCloseable` style in https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/681. However, some resource managers require an intermediate interface (e.g JDBC) so I have left them as is.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/AbstractPipelineLauncher.java:
```diff
@@ -0,0 +1,281 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.it.common;
```
"common" is not a very great module - it ends up just having everything in it. And the code in this file is for Dataflow, not common.

pranavbhandari24 replied:
I agree. I moved some of the files in common into separate modules. We can rename the `common` module to something like `core` if that's better.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/AbstractPipelineLauncher.java:
```diff
@@ -0,0 +1,281 @@
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
```
Other files use a Google `Pattern` class, but here the one that comes with Java is used.

pranavbhandari24 replied:
Fixed in https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/710.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/PipelineLauncher.java:
```diff
@@ -0,0 +1,422 @@
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableMap;
+import org.apache.beam.vendor.guava.v26_0_jre.com.google.common.collect.ImmutableSet;
+
+/** Client for working with Cloud Dataflow. */
```
Can this framework support running pipelines more generally, or just Dataflow?

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/PipelineLauncher.java:
```diff
@@ -0,0 +1,422 @@
+  }
+
+  /** Enum representing known Dataflow job states. */
+  enum JobState {
```
Aren't these already in the SDK?

pranavbhandari24 replied:
Not all of them. These represent the `JobState` values for dataflow. States like `DRAINING` are not in the SDK?

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/PipelineLauncher.java:
```diff
@@ -0,0 +1,422 @@
+    private static final String DATAFLOW_PREFIX = "JOB_STATE_";
+
+    /** States that indicate the job is getting ready to run. */
+    public static final ImmutableSet<JobState> PENDING_STATES = ImmutableSet.of(PENDING, QUEUED);
```
This stuff all seems like it would already be in the DataflowRunner code base, and that is the best place to add it.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/PipelineLauncher.java:
```diff
@@ -0,0 +1,422 @@
+        return jobName;
+      }
+
+      @Nullable
```
Use checkerframework `@Nullable` annotation and annotate the return type, not the overall method.

Do this throughout the file, and all the new code.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/PipelineOperator.java:
```diff
@@ -0,0 +1,322 @@
+   * @return the result, which will be {@link Result#LAUNCH_FINISHED}, {@link Result#LAUNCH_FAILED}
+   *     or {@link Result#TIMEOUT}
+   */
+  @SuppressWarnings("rawtypes")
```
Prefer to suppress this on a single line that creates the `Supplier[]` and assigns it to a variable with a proper type like `Supplier<Boolean>[]`

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/conditions/ConditionCheck.java:
```diff
@@ -0,0 +1,84 @@
+package org.apache.beam.it.common.conditions;
+
+import java.util.function.Supplier;
+import javax.annotation.Nullable;
```
use checker annotations

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/conditions/package-info.java:
```diff
@@ -0,0 +1,18 @@
+ */
+
+/** Package that contains reusable conditions. */
+package org.apache.beam.it.common.conditions;
```
Don't make subpackages of `common`. Common is already "misc" so you can just put `conditions` at the top level.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/logging/LogStrings.java:
```diff
@@ -0,0 +1,64 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.it.common.logging;
```
Same. Just `org.apache.beam.it.logging`

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/matchers/LaunchInfoSubject.java:
```diff
@@ -0,0 +1,59 @@
+  private final LaunchInfo actual;
+
+  private LaunchInfoSubject(FailureMetadata metadata, LaunchInfo actual) {
+    super(metadata, actual);
```
Using `super` constructor just to initialize fields is an anti-pattern. It seems to be built into Truth, which is a bummer. Is this the intended use of `Subject`? Is there a way to use composition instead of inheritance to more cleanly separate this class?

pranavbhandari24 replied:
Yeah, it is built into Truth. Inheritance is the recommended way of using `Subject`. See https://truth.dev/extension.html

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/matchers/ListAccumulator.java:
```diff
@@ -0,0 +1,61 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.it.common.matchers;
```
just `org.apache.beam.it.matchers`

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/matchers/package-info.java:
```diff
@@ -0,0 +1,18 @@
+ */
+
+/** Package for Truth matchers / subjects to have reusable assertions. */
+package org.apache.beam.it.common.matchers;
```
Since it is specifically for truth matchers, perhaps `org.apache.beam.it.truthmatchers` or something. It is fine as-is but imagine we had built the framework for Hamcrest and then Truth came out and we wanted to also support it. Separating modules according to the dependencies they pull in is often a good idea.

kennknowles commented on it/common/src/main/java/org/apache/beam/it/common/package-info.java:
```diff
@@ -0,0 +1,18 @@
+ */
+
+/** Package for managing common ResourceManager resources within integration tests. */
+package org.apache.beam.it.common;
```
just drop the `common`

pranavbhandari24 replied:
Addressed in above comment.

kennknowles commented on it/mongodb/build.gradle:
```diff
@@ -0,0 +1,39 @@
+    implementation library.java.slf4j_api
+    //implementation 'org.mongodb:mongodb-driver-sync:3.12.10'
+    implementation library.java.mongo_java_driver
+    implementation 'com.google.re2j:re2j:1.6'
```
same comment here - why use the Google library? should at least have a good reason and comment

pranavbhandari24 replied:
Fixed in https://github.com/GoogleCloudPlatform/DataflowTemplates/pull/710.

kennknowles commented on it/mongodb/src/main/java/org/apache/beam/it/mongodb/DefaultMongoDBResourceManager.java:
```diff
@@ -0,0 +1,280 @@
+  public static final class Builder
+      extends TestContainerResourceManager.Builder<DefaultMongoDBResourceManager> {
+
+    @SuppressWarnings("initialization.fields.uninitialized")
```
Why? Based on the API here, you should make this nullable, and check for null when you use it.

pranavbhandari24 replied:
Will address after code drop. Opened https://github.com/apache/beam/issues/27438

kennknowles commented on it/mongodb/src/test/java/org/apache/beam/it/mongodb/DefaultMongoDBResourceManagerTest.java:
```diff
@@ -0,0 +1,310 @@
+
+  @Test
+  public void testGetDatabaseNameShouldReturnCorrectValue() {
+    assertThat(testManager.getDatabaseName()).matches(TEST_ID + "-\\d{8}-\\d{6}-\\d{6}");
```
All the setup for the test is outside of the test case. You should inline it so that what is being tested is clear. Currently it basically just tests the setup.

pranavbhandari24 replied:
Refactored the test class to make it more clear. Thanks!

kennknowles commented on it/mongodb/src/test/java/org/apache/beam/it/mongodb/DefaultMongoDBResourceManagerTest.java:
```diff
@@ -0,0 +1,310 @@
+  @Mock private MongoDatabase database;
+  @Mock private MongoCollection<Document> collection;
+  @Mock private MongoCursor<String> collectionNames;
+  @Mock private MongoDBContainer container;
```
Overall, I'm concerned that the amount of mocking means that this file may create vacuous tests of our assumptions of how things behave, but do not test how they actually might behave. Is there any other way to have a lightweight fake?

pranavbhandari24 replied:
Thanks, I agree. I have reduced the amount of mocks in test classes.

kennknowles reviewed (changes requested):
Another round of comments just as I scrolled through a bit more of the code.

kennknowles commented:
waiting on author

pranavbhandari24 commented:
I'm working on addressing your comments and making the required changes to https://github.com/GoogleCloudPlatform/DataflowTemplates/tree/main/it to maintain one source of truth.

kennknowles commented:
OK let me know when I should make another pass.

damccorm commented:
stop reviewer notifications

pranavbhandari24 commented:
@kennknowles I have addressed your comments. The PR is ready for another pass.

pranavbhandari24 commented:
R: @Abacn, @johnjcasey

johnjcasey commented on it/testcontainers/build.gradle:
```diff
@@ -0,0 +1,37 @@
+    implementation project(path: ":it:common", configuration: "shadow")
+    implementation library.java.testcontainers_base
+
+    testImplementation 'com.google.truth:truth:1.0.1'
```
This should be done as a library, similar to our other libraries

johnjcasey commented on it/common/build.gradle:
```diff
@@ -0,0 +1,48 @@
+    implementation library.java.vendored_guava_26_0_jre
+    implementation library.java.slf4j_api
+    implementation library.java.commons_lang3
+    implementation 'dev.failsafe:failsafe:3.3.0'
```
please do this as a library in the plugin

johnjcasey commented on it/common/src/main/java/org/apache/beam/it/common/TestProperties.java:
```diff
@@ -0,0 +1,294 @@
+      // if on Compute Engine, return default credentials.
+      GoogleCredentials applicationDefault = ServiceAccountCredentials.getApplicationDefault();
+      try {
+        if (applicationDefault instanceof ComputeEngineCredentials) {
```
this looks quite strange, what throws an exception here?

johnjcasey commented on it/common/src/main/java/org/apache/beam/it/common/utils/ResourceManagerUtils.java:
```diff
@@ -0,0 +1,194 @@
+    }
+  }
+
+  public static String generatePassword(
```
This should have some commentary on how it should be used. This is not cryptographically secure, but is probably fine for temporary test envs.

johnjcasey commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/CassandraResourceManager.java:
```diff
@@ -0,0 +1,310 @@
+  private static final String DEFAULT_CASSANDRA_CONTAINER_TAG = "4.1.0";
+
+  // 27017 is the default port that Cassandra is configured to listen on
+  private static final int CASSANDRA_INTERNAL_PORT = 9042;
```
the comment and the variable don't allign

johnjcasey reviewed (changes requested):
This can be merged as is, but these should be cleaned up

kennknowles commented on it/cassandra/gradle.properties:
```diff
@@ -0,0 +1,18 @@
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+enableCheckerFramework=false
```
remove

kennknowles commented on it/cassandra/src/main/java/org/apache/beam/it/cassandra/CassandraResourceManager.java:
```diff
@@ -0,0 +1,310 @@
+ *
+ * <p>The class is thread-safe.
+ */
+public class CassandraResourceManager extends TestContainerResourceManager<GenericContainer<?>>
```
Could we merge the framework and the Cassandra LT in separate pull requests? I want to get this in ASAP but 25k lines with a lot of error checking disabled is a bit of a problem.

kennknowles commented on it/common/gradle.properties:
```diff
@@ -0,0 +1,18 @@
+#  See the License for the specific language governing permissions and
+# limitations under the License.
+################################################################################
+enableCheckerFramework=false
```
remove

kennknowles reviewed (changes requested):
Last time if I recall most of my concerns were disabling type checking. It seems like it is even more disabled now? And additional modules have been added to make the PR even bigger. Please just sync over the framework and do not disable checkerframework for whole directories. If it is too difficult to get specific classes to typecheck, you can disable them at a class level and comment next to the `@SuppressWarnings` with a GitHub Issue to fix it. We really should not be adding new suppressions to the repository. I recommend adding checker to the template repository as well, to avoid introducing more bugs.

kennknowles commented:
BTW the reason I am suggesting the split: this code clearly needs review & revision to match the expectations of the Beam repo. That will be much easier to do in pieces.

 - We can merge the overall framework with TODOs in place / P1s filed to fix the disabled checks
 - Then we could do thorough reviews one at a time of the test modules and get them fixed up before merge
